src/target/riscv/riscv-013.c

```diff
@@ -2811,6 +2818,324 @@ void riscv013_fill_dmi_nop_u64(struct target *target, char *buf)
 	buf_set_u64((unsigned char *)buf, DTM_DMI_ADDRESS_OFFSET, info->abits, 0);
 }
 
+static int get_max_sbaccess(struct target *target)
+{
+	RISCV013_INFO(info);
+
+	uint32_t sbaccess128 = get_field(info->sbcs, DMI_SBCS_SBACCESS128);
+	uint32_t sbaccess64	= get_field(info->sbcs, DMI_SBCS_SBACCESS64);
+	uint32_t sbaccess32	= get_field(info->sbcs, DMI_SBCS_SBACCESS32);
+	uint32_t sbaccess16	= get_field(info->sbcs, DMI_SBCS_SBACCESS16);
+	uint32_t sbaccess8 = get_field(info->sbcs, DMI_SBCS_SBACCESS8);
+
+	if (sbaccess128)
+		return 4;
+	else if (sbaccess64)
+		return 3;
+	else if (sbaccess32)
+		return 2;
+	else if (sbaccess16)
+		return 1;
+	else if (sbaccess8)
+		return 0;
+	else
+		return -1;
+}
+
+static uint32_t get_num_sbdata_regs(struct target *target)
+{
+	RISCV013_INFO(info);
+
+	uint32_t sbaccess128 = get_field(info->sbcs, DMI_SBCS_SBACCESS128);
+	uint32_t sbaccess64	= get_field(info->sbcs, DMI_SBCS_SBACCESS64);
+
+	if (sbaccess128)
+		return 4;
+	else if (sbaccess64)
+		return 2;
+	else
+		return 1;
+}
+
+static int riscv013_test_sba_config_reg(struct target *target,
+		target_addr_t legal_address, uint32_t num_words,
+		target_addr_t illegal_address, bool run_sbbusyerror_test)
+{
+	LOG_INFO("Testing System Bus Access as defined by RISC-V Debug Spec v0.13");
+
+	uint32_t rd_val;
+	uint32_t sbcs_orig;
+	dmi_read(target, &sbcs_orig, DMI_SBCS);
+
+	uint32_t sbcs = sbcs_orig;
+	bool test_passed;
+
+	int max_sbaccess = get_max_sbaccess(target);
+
+	if (max_sbaccess == -1) {
+		LOG_ERROR("System Bus Access not supported in this config.");
+		return ERROR_FAIL;
+	}
+
+	if (get_field(sbcs, DMI_SBCS_SBVERSION) != 1) {
+		LOG_ERROR("System Bus Access unsupported SBVERSION (0). Only version 1 is supported.");
+		return ERROR_FAIL;
+	}
+
+	uint32_t num_sbdata_regs = get_num_sbdata_regs(target);
+
+	/* Test 1: Simple write/read test */
+	test_passed = true;
+	sbcs = set_field(sbcs_orig, DMI_SBCS_SBAUTOINCREMENT, 0);
+	dmi_write(target, DMI_SBCS, sbcs);
+
+	uint32_t test_patterns[4] = {0xdeadbeef, 0xfeedbabe, 0x12345678, 0x08675309};
+	for (uint32_t sbaccess = 0; sbaccess <= (uint32_t)max_sbaccess; sbaccess++) {
+		sbcs = set_field(sbcs, DMI_SBCS_SBACCESS, sbaccess);
+		dmi_write(target, DMI_SBCS, sbcs);
+
+		uint32_t compare_mask = (sbaccess == 0) ? 0xff :
+														(sbaccess == 1) ? 0xffff : 0xffffffff;
```

Comment:
whats going on with this whitespace?

Reply:
Not sure... looks completely different in my local directory. I'll try reformatting